wandb/sdk/lib/paths.py

```diff
@@ -19,3 +20,139 @@
+        if callable(result):
+
+            @wraps(result)
+            def wrapper(*args: Any, **kwargs: Any) -> Any:
+                inner_result = result(*args, **kwargs)
+                if isinstance(inner_result, PurePosixPath):
+                    return LogicalPath(inner_result)
+                return inner_result
+
+            return wrapper
+        return result
+
+    def __truediv__(self, other: StrPath) -> "LogicalPath":
+        """Act like a PurePosixPath for the / operator, but return a LogicalPath."""
+        return LogicalPath(self.to_path() / other)
+
+
+PROHIBITED_CHARS = r'<>:"|?*'
+RESERVED_NAMES = (
+    ["CON", "PRN", "AUX", "NUL"]
+    + ["COM1", "COM2", "COM3", "COM4", "COM5", "COM6", "COM7", "COM8", "COM9"]
+    + ["LPT1", "LPT2", "LPT3", "LPT4", "LPT5", "LPT6", "LPT7", "LPT8", "LPT9"]
+)
+
+
+def sanitize_path(path: StrPath) -> PurePosixPath:
+    """Convert a path to a Path that can be used as a relative path on any platform.
+
+    This destructively modifies the path to be POSIX-style, removing anchors, converting
+    "legal" backslashes to forward slashes, and removing or replacing all characters
+    that might be allowed locally but are not allowed universally.
+    """
+    # Convert any non-pathlib object to a native PurePath.
+    if not isinstance(path, PurePath) and hasattr(path, "__fspath__"):
+        path = path.__fspath__()
+    if isinstance(path, bytes):
+        path = os.fsdecode(path)
+    if isinstance(path, str):
+        path = PurePath(path)
+
+    # If absolute, make relative to the root/drive. Handling pathological inputs means
+    # we can't guarantee a path has a valid representation on the current platform.
+    try:
+        pure_path = PurePath(path)
+        path = pure_path.relative_to(pure_path.anchor)
+    except ValueError:
+        pure_path = PurePosixPath(path)
+        path = pure_path.relative_to(pure_path.anchor)
+
+    # Remove unprintable characters.
+    path_str = "".join(c for c in path.as_posix() if c.isprintable())
+
+    # Replace all backslashes with forward slashes.
+    path = PurePosixPath(path_str.replace("\\", "/"))
+
+    # We do this again because the previous steps may have introduced a new root.
+    if path.anchor:
+        path = path.relative_to(path.anchor)
+
+    # Replace characters not allowed in Windows filenames.
+    path_str = "".join(c if c not in PROHIBITED_CHARS else "_" for c in path.as_posix())
+
+    # Strip trailing dots and spaces (another Windows requirement).
+    # Also normalize by eliminating trailing slashes.
+    path_str = path_str.rstrip(" ./")
+
+    # Alter any path parts that are reserved on Windows.
+    parts = list(PurePosixPath(path_str).parts)
+    parts = [part if part not in RESERVED_NAMES else f"_{part}" for part in parts]
+    posix_path = PurePosixPath(*parts)
+    if posix_path.name.split(".")[0] in RESERVED_NAMES:
+        posix_path = posix_path.with_name(f"_{posix_path.name}")
```

Comment:
Light googling suggests Windows paths are case-insensitive. Should we throw in some `.upper()`s?

```suggestion
    parts = list(PurePosixPath(path_str).parts)
    parts = [part if part.upper() not in RESERVED_NAMES else f"_{part}" for part in parts]
    posix_path = PurePosixPath(*parts)
    if posix_path.name.split(".")[0].upper() in RESERVED_NAMES:
        posix_path = posix_path.with_name(f"_{posix_path.name}")
```

Reply:
I'm actually going to rip all the sanitize code out. I don't have immediate plans for it, and it's bad form to check in code that I don't have an explicit use for. Also, it was fiendishly complex to test and get right, so it probably still isn't.